### PR TITLE
Fix regression where renaming a CodeElement didn't update node keys

### LINKED ISSUE #31735: Can't save .cs file after changing the Name property of a Button in Winforms property grid
**Version Used**: 
Dev16 (d16.0stg branch), works fine in VS 2017
This is customer feedback from the preview channel - 
https://developercommunity.visualstudio.com/content/problem/400702/cant-save-cs-file-after-changing-the-name-attribut.html

**Steps to Reproduce**:
1 Create a Winforms application.
2 Drag and drop a Button control to Form1 designer.
3 Rename the Name property for this Button control.
4 Save the Form1.cs file.

**Expected Behavior**:
file is saved

**Actual Behavior**:
error box 
image

```
Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.AbstractCodeModelService.LookupNode(Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.SyntaxNodeKey nodeKey, Microsoft.CodeAnalysis.SyntaxTree syntaxTree) Line 153	C#
  Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.InternalElements.AbstractKeyedCodeElement.LookupNode() Line 55	C#
  Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.InternalElements.AbstractCodeElement.GetName() Line 106	C#
  Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.InternalElements.AbstractKeyedCodeElement.GetName() Line 109	C#
  Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel.InternalElements.AbstractCodeElement.Name.get() Line 117	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
  Microsoft.VisualStudio.Design.dll!Microsoft.VisualStudio.Design.Serialization.CodeDom.VSCodeDomGenerator.GenerateCodeFromType(System.CodeDom.CodeTypeDeclaration e, System.IO.TextWriter w, System.CodeDom.Compiler.CodeGeneratorOptions o)	Unknown
```

### PR DESCRIPTION
**Pull request by jasonmalinowski**
In a larger refactoring I had commented some code out in NodeKeyValidation.AddProject. This meant that when we did a refactor rename of a CodeElement, we wouldn't end up updating node keys for all the existing CodeElements out there. The expectation is CodeElements are still valid. Uncommenting the code and bringing it back inline with the current design makes everything work again.

Fixes #31735.

<details><summary>Ask Mode template not completed</summary>

**Customer scenario**

Change the name of a control in a Win Forms project. Once you try to save it, you get an error saying your file can't be saved. At this point, all you can do is close the file without saving and try again.

**Bugs this fixes:** 

GitHub Bug: #31735
Feedback link: https://developercommunity.visualstudio.com/content/problem/400702/cant-save-cs-file-after-changing-the-name-attribut.html

**Workarounds, if any**

You'd have to perform the rename manually in the codebehind of the file. This isn't an obvious workaround (the error the customer gives does not give any hint which operation caused you to get into this state), and is also fairly risky since we normally don't want customers tinkering in the codebehind.

**Risk**

Low: this is fixing one particular method which is only used in the rename path for CodeModel elements.

**Performance impact**

None.

**Is this a regression from a previous update?**

Yes. This was broken in Dev16.0 Preview 1.

**Root cause analysis:**

This was broken as a part of a very large refactoring; some code wasn't fixed up correctly and this slipped through code review. There weren't any unit tests covering this, and any integration tests must have since been disabled. Unit tests have now been written to cover this, which were confirmed to correctly catch the bug before the fix was checked in.

**How was the bug found?**

Many customer reports: https://developercommunity.visualstudio.com/content/problem/400702/cant-save-cs-file-after-changing-the-name-attribut.html

</details>

### CODE REVIEW

File: src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs [hunk 1]
```diff
@@ -488,30 +488,15 @@ protected EnvDTE80.CodeAttributeArgument CreateInternalCodeAttributeArgument(Cod
...
-        public void Rename(ISymbol symbol, string newName, Solution solution)
+        public void Rename(ISymbol symbol, string newName, Workspace workspace, ProjectCodeModelFactory projectCodeModelFactory)
         {
-            // TODO: (tomescht) make this testable through unit tests.
-            // Right now we have to go through the project system to find all
-            // the outstanding CodeElements which might be affected by the
-            // rename. This is silly. This functionality should be moved down
-            // into the service layer.
-
-            var workspace = solution.Workspace as VisualStudioWorkspace;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs]: I'm not including this check; it's unclear to me how somebody would get a CodeElement not backed by a VisualStudioWorkspace to begin with. And also nothing here actually used the VisualStudioWorkspace-ness of it in any way.

File: src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb [hunk 2]
```diff
@@ -49,19 +49,22 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.CodeModel
...
-                Dim mockTextManagerAdapter = New MockTextManagerAdapter()
+                Dim projectCodeModel = DirectCast(state.ProjectCodeModelFactory.CreateProjectCodeModel(project.Id, Nothing), ProjectCodeModel)
 
-                For Each documentId In project.DocumentIds
+                For Each document In project.Documents
                     ' Note that a parent is not specified below. In Visual Studio, this would normally be an EnvDTE.Project instance.
-                    Dim fcm = FileCodeModel.Create(state, parent:=Nothing, documentId:=documentId, textManagerAdapter:=mockTextManagerAdapter)
-                    mockVisualStudioWorkspace.SetFileCodeModel(documentId, fcm)
+                    Dim fcm = projectCodeModel.GetOrCreateFileCodeModel(document.FilePath, parent:=Nothing)
+                    fcm.Object.TextManagerAdapter = New MockTextManagerAdapter()
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: I will totally admit assigning this here is hacky, but it's the best I can do without a much larger refactoring if we want to get this in for Preview 2.

File: src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs [hunk 3]
```diff
@@ -488,30 +488,15 @@ protected EnvDTE80.CodeAttributeArgument CreateInternalCodeAttributeArgument(Cod
 
         public abstract string GetFullyQualifiedName(string name, int position, SemanticModel semanticModel);
 
-        public void Rename(ISymbol symbol, string newName, Solution solution)
+        public void Rename(ISymbol symbol, string newName, Workspace workspace, ProjectCodeModelFactory projectCodeModelFactory)
         {
-            // TODO: (tomescht) make this testable through unit tests.
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs]: I'm adding tests in this PR.

File: src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs (hunk 1 above)
Comment (DustinCampbell) [reply src/VisualStudio/Core/Impl/CodeModel/AbstractCodeModelService.cs]: Fair enough

File: src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb (hunk 2 above)
Comment (DustinCampbell) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: That's fine. I was equally conflicted as I read through the rest of the code, but hadn't said anything yet. Glad we're of the same mind. :smile:

File: src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb [hunk 4]
```diff
@@ -91,7 +94,7 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.CodeModel
                     Return Me._componentModel
                 End If
 
-                Throw New NotImplementedException()
+                Throw New NotImplementedException("No service exists for " + serviceType.FullName)
```
Comment (DustinCampbell) [inline src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: nit: interpolated string.

File: src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: I *did* try to thread it through and ended up in a much bigger mess...

Comment (sharwell) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: Can we at least change it to the `Testaccessor` pattern, and make the setter of the main property `private`?

Comment (jasonmalinowski) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: Given this is being considered for Preview 2, I'm not going to touch this _right now_. I think that's a great suggestion and if it merges in this form I'll dig into this further.

The other approach I think makes more sense is just exposing this either as a MEF or workspace service or IServiceProvider service, or something else. The core operation needed is "give me a EnvDTE.TextPoint" which in VS has to be done through the shell. For tests, we just use a mock. Any of our standard dependency injection patterns would then mean we avoid the need of this having to do "special" stuff in this way.

Comment (jasonmalinowski) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: And oh: if this is rejected for Preview 2, then I'll give another crack at this for Preview 3.

Comment (jasonmalinowski) [reply src/VisualStudio/TestUtilities2/CodeModel/CodeModelTestHelpers.vb]: ...and of course as I write that, it's approved for Preview 2. :smile:

**Pull request merged**